### PR TITLE
save category mappings to new Category MappingsTable

### PR DESCRIPTION
**Pull request by Sravanksk**

### CODE REVIEW

File: src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts [hunk 1]
```diff
@@ -165,23 +136,19 @@ export class CategoryMappingsDialogComponent implements OnInit {
...
     that.setupSageIntacctExpenseTypeWatchers();
-    this.setupSageIntacctCCCAccountWatchers();
   }
 
   getAttributesFilteredByConfig() {
     const that = this;
     const attributes = [];
 
-    if (that.configuration.reimbursable_expenses_object === 'EXPENSE_REPORT') {
+    if (that.configuration.reimbursable_expenses_object === 'EXPENSE_REPORT' || (that.showSeparateCCCField() && that.configuration.corporate_credit_card_expenses_object === 'EXPENSE_REPORT')) {
```
Comment (ashwin1111) [inline src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts:145]: Why is this added? `|| (that.showSeparateCCCField() && that.configuration.corporate_credit_card_expenses_object === 'EXPENSE_REPORT')`

File: src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts [hunk 2]
```diff
@@ -201,21 +168,16 @@ export class CategoryMappingsDialogComponent implements OnInit {
...
-      const sageIntacctExpenseTypes = that.configuration.reimbursable_expenses_object === 'EXPENSE_REPORT' && that.editMapping ? that.sageIntacctExpenseTypes.filter(siExpTypeObj => siExpTypeObj.value === that.data.rowElement.si_value)[0] : '';
-      const sageIntacctCCCAccount = that.showSeparateCCCField() && that.editMapping ? that.sageIntacctCCCAccounts.filter(cccObj => cccObj.value === that.data.rowElement.ccc_value)[0] : '';
+      const fyleCategory = that.editMapping ? that.fyleCategories.filter(category => category.value === that.data.categoryMappingRow.source_category.value)[0] : '';
+      const sageIntacctAccount = that.editMapping ? that.sageIntacctAccounts.filter(siAccObj => that.data.categoryMappingRow.destination_account && siAccObj.value === that.data.categoryMappingRow.destination_account.value)[0] : '';
+      const sageIntacctExpenseTypes = that.configuration.reimbursable_expenses_object === 'EXPENSE_REPORT' && that.editMapping ? that.sageIntacctExpenseTypes.filter(siExpTypeObj => that.data.categoryMappingRow.destination_expense_head && siExpTypeObj.value === that.data.categoryMappingRow.destination_expense_head.value)[0] : '';
```
Comment (ashwin1111) [inline src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts:175]: sageIntacctExpenseType

File: src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts [hunk 3]
```diff
@@ -201,21 +168,16 @@ export class CategoryMappingsDialogComponent implements OnInit {
...
 
       that.form = that.formBuilder.group({
         fyleCategory: [fyleCategory, Validators.compose([Validators.required, that.forbiddenSelectionValidator(that.fyleCategories)])],
-        sageIntacctAccount: [sageIntacctAccount, that.configuration.reimbursable_expenses_object === 'BILL' ? that.forbiddenSelectionValidator(that.sageIntacctAccounts) : null],
-        sageIntacctExpenseTypes: [sageIntacctExpenseTypes, that.configuration.reimbursable_expenses_object === 'EXPENSE_REPORT' ? that.forbiddenSelectionValidator(that.sageIntacctExpenseTypes) : null],
-        sageIntacctCCCAccount: [sageIntacctCCCAccount, that.showSeparateCCCField() ? that.forbiddenSelectionValidator(that.sageIntacctCCCAccounts) : null],
+        sageIntacctAccount: [sageIntacctAccount],
```
Comment (ashwin1111) [inline src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts:179]: Why did we ignore `forbiddenSelectionValidator`?

File: src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts [hunk 4]
```diff
@@ -201,21 +168,16 @@ export class CategoryMappingsDialogComponent implements OnInit {
...
+        sageIntacctExpenseTypes: [sageIntacctExpenseTypes],
```
Comment (ashwin1111) [inline src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts:180]: Why did we ignore `forbiddenSelectionValidator`?

File: src/app/si/settings/category-mappings/category-mappings.component.html [hunk 5]
```diff
@@ -46,14 +46,14 @@
           {{ configuration.reimbursable_expenses_object === 'EXPENSE_REPORT' ? 'Expense Type' : 'Account' }}
         </th>
         <td mat-cell *matCellDef="let element">
-          {{element.si_value}}
+          {{ configuration?.reimbursable_expenses_object == 'EXPENSE_REPORT' ? element.destination_expense_head?.value : element.destination_account?.value }}
```
Comment (ashwin1111) [inline src/app/si/settings/category-mappings/category-mappings.component.html:49]: ===

File: src/app/si/settings/category-mappings/category-mappings.component.html [hunk 6]
```diff
@@ -46,14 +46,14 @@
...
         </td>
       </ng-container>
 
       <ng-container matColumnDef="ccc">
         <th mat-header-cell *matHeaderCellDef> Credit Card Expense Account </th>
         <td mat-cell *matCellDef="let element" >
-          {{element.ccc_value}}
+          {{ configuration?.corporate_credit_card_expenses_object == 'EXPENSE_REPORT' ? element.destination_expense_head?.value : element.destination_account?.value }}
```
Comment (ashwin1111) [inline src/app/si/settings/category-mappings/category-mappings.component.html:56]: ===

Comment (ashwin1111) [reply src/app/si/settings/category-mappings/category-mappings.component.html:56]: don't think this condition is required, we never show expense head in ccc row

File: src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts (hunk 3 above)
Comment (Sravanksk) [reply src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts:179]: since we are not checking based on this condition -> `showSeparateCCCField()`, `forbiddenSelectionValidator` is not needed here

File: src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts (hunk 4 above)
Comment (Sravanksk) [reply src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts:180]: same

File: src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts (hunk 3 above)
Comment (ashwin1111) [reply src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts:179]: `forbiddenSelectionValidator` helps in identifying invalid search terms.
Screenshot 2022-07-26 at 5 15 55 PM

Were you able to achieve this without `forbiddenSelectionValidator`?

File: src/app/si/settings/category-mappings/category-mappings.component.html (hunk 6 above)
Comment (Sravanksk) [reply src/app/si/settings/category-mappings/category-mappings.component.html:56]: yes makes sense

File: src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts (hunk 4 above)
Comment (ashwin1111) [reply src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts:180]: https://github.com/fylein/fyle-intacct-app/pull/102#discussion_r929864498

File: src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts (hunk 3 above)
Comment (Sravanksk) [reply src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts:179]: let me check it once 😅

Comment (Sravanksk) [reply src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts:179]: adding the validator now, seems like it is not working

File: src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts (hunk 1 above)
Comment (Sravanksk) [reply src/app/si/settings/category-mappings/category-mappings-dialog/category-mappings-dialog.component.ts:145]: checked the possible config combinations in sage-intacct, this check is not needed ☮️

**Review by ashwin1111 — APPROVED** [review]
